pkg/mic/mic.go

```diff
@@ -1317,3 +1327,117 @@ func (c *Client) checkIfIdentityImmutable(id string) bool {
 	}
 	return false
 }
+
+// generateIdentityAssignmentState generates the current and desired state of each node's identity
+// assignments based on an existing list of AzureAssignedIdentity as the source of truth.
+func (c *Client) generateIdentityAssignmentState() (currentState map[string]map[string]bool, desiredState map[string][]string, isVMSSMap map[string]bool, err error) {
+	type nodeMetadata struct {
+		nodeNameOnAzure string
+		isVMSS          bool
+	}
+
+	assignedIDs, err := c.CRDClient.ListAssignedIDs()
+	if err != nil {
+		return nil, nil, nil, fmt.Errorf("failed to list AzureAssignedIdentities, error: %+v", err)
+	}
+
+	nodeMetadataCache := make(map[string]nodeMetadata)
+	isVMSSMap = make(map[string]bool)
+	currentState = make(map[string]map[string]bool)
+	desiredState = make(map[string][]string)
+	for _, assignedID := range *assignedIDs {
+		if _, ok := nodeMetadataCache[assignedID.Spec.NodeName]; !ok {
+			node, err := c.NodeClient.Get(assignedID.Spec.NodeName)
+			if err != nil {
+				return nil, nil, nil, fmt.Errorf("failed to get node %s, error %+v", assignedID.Spec.NodeName, err)
+			}
+
+			nodeNameOnAzure, isVMSS, err := isVMSS(node)
```

Comment:
Let's rename this var to just `nodeName` as at this point that's all it represents